lib/spack/spack/bootstrap.py

```diff
@@ -543,7 +554,16 @@ def ensure_executables_in_path_or_raise(executables, abstract_spec):
         b = _make_bootstrapper(current_config)
         try:
             if b.try_search_path(executables, abstract_spec):
-                return
+                # Additional environment variables needed
+                env_mods = spack.util.environment.EnvironmentModifications()
+                cmd = spack.util.executable.which(*executables)
+                concrete_spec = spack.store.db.query(abstract_spec, installed=True)[0]
```

Comment:
Why do we actually search for an abstract spec? After bootstrapping, we know the concrete spec, right? Can't we store its hash and use that once and for all?

Reply:
(basically like using a spack.lock for bootstrapped packages :p)

Reply:
> (basically like using a spack.lock for bootstrapped packages :p)

An environment doesn't have a concept of "fallback specs", so we can't use it directly. At least right now. We can't e.g. try to install a rhel5 clingo and if that fails fall back to build another thing.

Reply:
See dabf285 for a way to address this without doing a lot of refactoring (which will probably be subject for a subsequent PR).